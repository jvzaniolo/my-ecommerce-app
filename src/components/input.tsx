import {
  Input as ChakraInput,
  FormControl,
  FormErrorMessage,
  FormHelperText,
  FormLabel,
  type InputProps as ChakraInputProps,
} from '@chakra-ui/react'
import { forwardRef, type ForwardRefRenderFunction } from 'react'

interface InputProps extends ChakraInputProps {
  label?: string
  helper?: string
  error?: string
}

const _input: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  { id, isRequired, isInvalid, helper, error, label, ...rest },
  ref
) => {
  return (
    <FormControl isInvalid={isInvalid} isRequired={isRequired}>
      {label && <FormLabel htmlFor={id}>{label}</FormLabel>}
      <ChakraInput ref={ref} name={id} {...rest} />
      {error ? (
        <FormErrorMessage>{error}</FormErrorMessage>
      ) : (
        <FormHelperText>{helper}</FormHelperText>
      )}
    </FormControl>
  )
}

const Input = forwardRef(_input)

export default Input
