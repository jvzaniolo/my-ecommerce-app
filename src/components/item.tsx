import { useRouter } from 'next/router'
import { Heading, Img, Stack, Text } from '@chakra-ui/react'
import type { Item as ItemProps } from '~/types/item'

export function Item({ id, name, description, price, image }: ItemProps) {
  const { push } = useRouter()

  function handleClick() {
    push(`/items/${id}`)
  }

  return (
    <Stack as="button" cursor="pointer" onClick={handleClick}>
      <Img src={image} alt={name} w="40" />
      <Heading size="sm">{name}</Heading>
      <Text>{description}</Text>
      <Text>
        {new Intl.NumberFormat('en-US', {
          style: 'currency',
          currency: 'USD',
        }).format(price)}
      </Text>
    </Stack>
  )
}
