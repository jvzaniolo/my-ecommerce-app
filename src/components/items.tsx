import Link from 'next/link'
import useSWR from 'swr'
import { Flex, Heading, Img, Text } from '@chakra-ui/react'
import { fetcher } from '~/services/axios'
import { toUSCurrency } from '~/utils/format'
import type { Item as ItemType } from '~/types/item'

export function Items() {
  const { data: items } = useSWR<ItemType[]>('/items', fetcher)

  return (
    <Flex gap="4" flexWrap="wrap">
      {items && items.length > 0 ? (
        items?.map(item => (
          <Link key={item.id} href={`/items/${item.slug}`} passHref>
            <Flex as="a" direction="column" shadow="md" maxW="72">
              <Img src={item.image} alt={item.name} w="72" borderTopRadius="md" />
              <Flex
                p="3"
                gap="3"
                flex="1"
                direction="column"
                borderBottomRadius="md"
              >
                <Heading size="sm">{item.name}</Heading>
                <Text noOfLines={3}>{item.description}</Text>
                <Text mt="auto" fontWeight="bold">
                  {toUSCurrency(item.price)}
                </Text>
              </Flex>
            </Flex>
          </Link>
        ))
      ) : (
        <Text>There are no items yet.</Text>
      )}
    </Flex>
  )
}
