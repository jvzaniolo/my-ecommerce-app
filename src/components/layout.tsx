import type { ReactNode } from 'react'

import Link from 'next/link'
import { MdOutlineShoppingCart } from 'react-icons/md'
import {
  Container,
  Flex,
  Grid,
  GridItem,
  Heading,
  IconButton,
  Spacer,
  Text,
} from '@chakra-ui/react'

const Layout = ({ children }: { children: ReactNode }) => {
  return (
    <Grid h="100vh" templateRows={'auto 1fr auto'}>
      <GridItem
        as="header"
        bg="white"
        top="0"
        pos="sticky"
        shadow="base"
        zIndex="sticky"
      >
        <Container maxW="container.xl">
          <Flex p="2" align="center">
            <Link href="/" passHref>
              <Heading as="a" size="sm">
                My E-Commerce App
              </Heading>
            </Link>
            <Spacer />
            <Link href="/cart" passHref>
              <IconButton
                as="a"
                aria-label="cart"
                icon={<MdOutlineShoppingCart />}
              />
            </Link>
          </Flex>
        </Container>
      </GridItem>

      <GridItem as="main">
        <Container h="full" maxW="container.xl" p="4">
          {children}
        </Container>
      </GridItem>

      <GridItem as="footer" borderTop="1px solid" borderColor="blackAlpha.200">
        <Container py="10" maxW="container.xl" centerContent>
          <Text color="blackAlpha.800" size="sm">
            Made with ðŸ’œ by @jvzaniolo | Copyright Â© 2022
          </Text>
        </Container>
      </GridItem>
    </Grid>
  )
}

export default Layout
