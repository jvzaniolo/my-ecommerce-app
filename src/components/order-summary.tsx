import type { Item } from '~/types/item'

import Link from 'next/link'
import {
  Box,
  Button,
  Divider,
  Flex,
  Heading,
  Stack,
  Text,
} from '@chakra-ui/react'
import { toUSCurrency } from '~/utils/format'
import { useSWRConfig } from 'swr'

const OrderSummary = () => {
  const { cache } = useSWRConfig()

  const cart: Item[] = cache.get('/cart')

  const cartTotal = cart.reduce((acc, item) => {
    return acc + item.price * item.quantity
  }, 0)

  return (
    <Flex
      p="4"
      h="sm"
      flex="1"
      shadow="md"
      direction="column"
      borderRadius="md"
    >
      <Stack spacing="4">
        <Heading size="lg">Order Summary</Heading>
        {cart?.map(item => (
          <Box key={item.id}>
            <Text noOfLines={1}>{item.name}</Text>
            <Flex justify="space-between">
              <Text color="gray.500">Quantity: {item.quantity}</Text>
              <Text>{toUSCurrency(item.price * item.quantity)}</Text>
            </Flex>
          </Box>
        ))}
        <Divider />
        <Flex justify="space-between">
          <Text fontWeight="bold">Total:</Text>
          <Text fontWeight="bold">{toUSCurrency(cartTotal || 0)}</Text>
        </Flex>
      </Stack>

      <Link href="/checkout" passHref>
        <Button mt="auto" as="a" w="full">
          Checkout
        </Button>
      </Link>
    </Flex>
  )
}

export default OrderSummary
