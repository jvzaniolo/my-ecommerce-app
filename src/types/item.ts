export type Item = {
  id: number
  name: string
  description: string
  price: number
  image: string
  quantity: number
}
