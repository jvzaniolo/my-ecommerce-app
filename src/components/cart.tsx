import NextLink from 'next/link'
import Image from 'next/image'
import useSWR, { useSWRConfig } from 'swr'
import {
  Box,
  Breadcrumb,
  BreadcrumbItem,
  BreadcrumbLink,
  Button,
  Divider,
  Flex,
  Heading,
  Img,
  Stack,
  Text,
  Link,
} from '@chakra-ui/react'
import { toUSCurrency } from '~/utils/format'
import { Quantity } from '~/components/quantity'
import { api, fetcher } from '~/services/axios'
import { Item } from '~/types/item'

export function Cart() {
  const { mutate } = useSWRConfig()
  const { data: cart } = useSWR<Item[]>('/cart', fetcher)

  console.log({ cart })

  const cartTotal = cart?.reduce((acc, item) => {
    return acc + item.price * item.quantity
  }, 0)

  function onRemoveItem(id: string) {
    const filteredCart = cart?.filter(item => item.id !== id)

    mutate(
      '/cart',
      async () => {
        await api.delete(`/cart/${id}`)
        return filteredCart
      },
      {
        optimisticData: filteredCart,
        rollbackOnError: true,
        revalidate: false,
        populateCache: true,
      }
    )
  }

  function onChangeQuantity(id: string, quantity: number) {
    const cartWithUpdatedQuantity = cart?.map(item =>
      item.id === id ? { ...item, quantity } : item
    )

    mutate(
      '/cart',
      async () => {
        await api.patch(`/cart/${id}`, { quantity })
        return cartWithUpdatedQuantity
      },
      {
        optimisticData: cartWithUpdatedQuantity,
        rollbackOnError: true,
        revalidate: false,
        populateCache: true,
      }
    )
  }

  return (
    <>
      <Breadcrumb mt="3" mb="6">
        <BreadcrumbItem>
          <NextLink href="/" passHref>
            <BreadcrumbLink>Home</BreadcrumbLink>
          </NextLink>
        </BreadcrumbItem>
        <BreadcrumbItem isCurrentPage>
          <BreadcrumbLink>Cart</BreadcrumbLink>
        </BreadcrumbItem>
      </Breadcrumb>

      <Box>
        <Heading as="h1" size="xl">
          Cart
        </Heading>

        <Flex direction={['column', 'column', 'row']} gap="10" mt="4">
          <Stack as="ul" flex="2" spacing="4">
            {cart && cart.length > 0 ? (
              cart?.map(item => (
                <Flex
                  key={item.id}
                  as="li"
                  h="56"
                  shadow="md"
                  borderRightRadius="md"
                >
                  <Box pos="relative" flex="1">
                    <Img
                      as={Image}
                      layout="fill"
                      src={item.image}
                      alt={item.name}
                      borderLeftRadius="md"
                    />
                  </Box>
                  <Flex p="4" h="full" direction="column" gap="2" flex="2">
                    <NextLink href={`/items/${item.slug}`}>
                      <Link size="md" fontWeight="bold" fontSize="xl">
                        {item.name}
                      </Link>
                    </NextLink>

                    <Text noOfLines={3} title={item.description}>
                      {item.description}
                    </Text>
                    <Quantity
                      value={item.quantity}
                      onChange={quantity => onChangeQuantity(item.id, quantity)}
                      maxQuantity={item.stock}
                    />

                    <Flex justify="space-between" mt="auto">
                      <Text fontWeight="bold">{toUSCurrency(item.price)}</Text>
                      <Button
                        variant="link"
                        onClick={() => onRemoveItem(item.id)}
                      >
                        Remove
                      </Button>
                    </Flex>
                  </Flex>
                </Flex>
              ))
            ) : (
              <Text>Your cart is empty</Text>
            )}
          </Stack>

          <Flex
            p="4"
            h="sm"
            flex="1"
            shadow="md"
            direction="column"
            borderRadius="md"
          >
            <Stack spacing="4">
              <Heading size="lg">Order Summary</Heading>
              {cart?.map(item => (
                <Box key={item.id}>
                  <Text noOfLines={1}>{item.name}</Text>
                  <Flex justify="space-between">
                    <Text color="gray.500">Quantity: {item.quantity}</Text>
                    <Text>{toUSCurrency(item.price * item.quantity)}</Text>
                  </Flex>
                </Box>
              ))}
              <Divider />
              <Flex justify="space-between">
                <Text fontWeight="bold">Total:</Text>
                <Text fontWeight="bold">{toUSCurrency(cartTotal || 0)}</Text>
              </Flex>
            </Stack>

            <NextLink href="/checkout" passHref>
              <Button mt="auto" as="a" w="full">
                Checkout
              </Button>
            </NextLink>
          </Flex>
        </Flex>
      </Box>
    </>
  )
}
