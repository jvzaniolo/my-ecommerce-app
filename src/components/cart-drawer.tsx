import {
  AspectRatio,
  Button,
  Drawer,
  DrawerBody,
  DrawerCloseButton,
  DrawerContent,
  DrawerFooter,
  DrawerHeader,
  DrawerOverlay,
  Flex,
  Heading,
  Img,
  Stack,
  Text,
} from '@chakra-ui/react'
import { AxiosError } from 'axios'
import Image from 'next/image'
import Link from 'next/link'
import { useRouter } from 'next/router'
import { FC } from 'react'
import useSWR from 'swr'
import { useCartDrawer } from '~/contexts/cart-drawer'
import { optimisticDeleteItem, optimisticUpdateItemQuantity } from '~/lib/cart'
import { Cart } from '~/types'
import { fetcher } from '~/utils/axios'
import { toUSCurrency } from '~/utils/format'
import { Quantity } from './quantity'

type CartDrawerProps = {
  isOpen: boolean
  onClose: () => void
}

export const CartDrawer: FC<CartDrawerProps> = ({ isOpen, onClose }) => {
  const { data: cart, error } = useSWR<Cart, AxiosError>('/api/cart', fetcher)
  const { onCloseCartDrawer } = useCartDrawer()
  const router = useRouter()

  const cartTotal = cart
    ? cart.items.reduce((acc, item) => {
        return acc + item.product.price * item.quantity
      }, 0)
    : 0

  if (error) return <>{error.message}</>

  return (
    <Drawer size="sm" isOpen={isOpen} placement="right" onClose={onClose}>
      <DrawerOverlay />
      <DrawerContent>
        <DrawerCloseButton />
        <DrawerHeader shadow="md">Cart</DrawerHeader>

        <DrawerBody>
          <Stack spacing="8">
            {cart && cart.items.length > 0 ? (
              cart.items.map(item => (
                <Flex key={item.id} gap="3">
                  <AspectRatio ratio={1} flex="1">
                    <Img
                      as={Image}
                      src={item.product.image_url}
                      alt={item.product.name}
                      layout="fill"
                      objectFit="cover"
                    />
                  </AspectRatio>
                  <Flex flex="2" direction="column" justify="space-between">
                    <Heading as="h3" size="md" fontWeight="medium">
                      {item.product.name}
                    </Heading>
                    <Flex justify="space-between">
                      <Text fontSize="lg" fontWeight="semibold">
                        {toUSCurrency(item.product.price * item.quantity)}
                      </Text>
                      <Button
                        variant="link"
                        onClick={() => {
                          optimisticDeleteItem(cart, item.id)
                        }}
                      >
                        Remove
                      </Button>
                    </Flex>
                    <Quantity
                      max={item.product.stock}
                      value={item.quantity}
                      onChange={value => {
                        optimisticUpdateItemQuantity(cart, item.id, value)
                      }}
                    />
                  </Flex>
                </Flex>
              ))
            ) : (
              <Text>Your cart is empty.</Text>
            )}
          </Stack>
        </DrawerBody>

        <DrawerFooter>
          <Link href="/checkout" passHref>
            <Button
              as="a"
              flex="1"
              colorScheme="purple"
              onClick={event => {
                event.preventDefault()
                onCloseCartDrawer()
                router.push('/checkout')
              }}
            >
              Checkout: {toUSCurrency(cartTotal)}
            </Button>
          </Link>
        </DrawerFooter>
      </DrawerContent>
    </Drawer>
  )
}
