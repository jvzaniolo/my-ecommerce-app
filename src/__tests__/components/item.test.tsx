import { render } from '@testing-library/react'
import { Item } from '~/components/item'
import type { Item as ItemType } from '~/types/item'

const item: ItemType = {
  id: '1',
  slug: 'slug',
  name: 'name',
  description: 'description',
  image: 'image',
  price: 20.0,
  quantity: 1,
}

describe('Item', () => {
  it('should render an item with initial props', () => {
    const { container } = render(<Item {...item} />)

    expect(container).toMatchSnapshot()
  })
})
