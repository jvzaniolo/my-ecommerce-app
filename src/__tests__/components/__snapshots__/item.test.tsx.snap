// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render an item with initial props 1`] = `
<div>
  <a
    class="css-b5xe08"
    href="/items/slug"
  >
    <img
      alt="name"
      class="chakra-image css-ydteuf"
      src="image"
    />
    <div
      class="css-1g8wnvq"
    >
      <h2
        class="chakra-heading css-0"
      >
        name
      </h2>
      <p
        class="chakra-text css-1cp6yu0"
      >
        description
      </p>
      <p
        class="chakra-text css-1hwfm6p"
      >
        $20.00
      </p>
    </div>
  </a>
</div>
`;
